[Console] Convert all non-autocomplete lib files to typescript

curq opened the pull request:
### Description

Converts ```lib/!autocomplete``` to typescript. Part of the https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3825

### Issues Resolved

Solves https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3827

### Testing the changes

No type errors present in affected files, except from imports of other non-ts files, will be fixed by other PRs from the same group of issues (https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4008, https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3826). 

### Check List

- [X] All tests pass
  - [X] `yarn test:jest`
  - [X] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [X] Update [CHANGELOG.md](./../CHANGELOG.md)
- [X] Commits are signed per the DCO using --signoff

ananzh commented on src/plugins/console/public/lib/osd/osd.ts:
```diff
@@ -36,69 +38,102 @@ import {
-import _ from 'lodash';
-
-import Api from './api';
+import { Api } from './api';
```
nice

ananzh commented on src/plugins/console/public/lib/osd/osd.ts:
```diff
@@ -36,69 +38,102 @@ import {
+  multiValued?: boolean | unknown
+) => SharedComponent;
+
+export interface ParametrizedComponentFactories {
```
I see this `ParametrizedComponentFactories` type is used in multiple places. Should we create a common folder or some type file to host it?

curq replied:
1. You right, we should move it, I will add it to types file that I created in my other PR. I forgot to mention the order of merging, it should first be https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4008, then https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148 and only after that this one.

ananzh replied:
I see. Then this PR is blocked by #4008 and #4148. Let's get back to this PR after them merged. @curq I will start reviewing these two.

curq replied:
I moved it to `types.ts` files that holds the common types

ananzh commented on src/plugins/console/public/lib/osd/osd.ts:
```diff
@@ -36,69 +38,102 @@ import {
-      return this[name];
+const parametrizedComponentFactories: ParametrizedComponentFactories = {
+  getComponent(name, parent, provideDefault) {
+    if ((this as any)[name]) {
```
`any` type seems not ideal. this is binding to ParametrizedComponentFactories context right?
should we do `getComponent(name: keyof ParametrizedComponentFactories ... )`?

curq replied:
I agree, I will try updating it after other 2 related PRs will be merged so that I could transfer the interface to the separate file and update it there.

curq replied:
I have updated this part and removed usage of `any`

ananzh commented on src/plugins/console/public/lib/osd/api.ts:
```diff
@@ -29,12 +29,14 @@
   globalsOnlyAutocompleteComponents,
   compileBodyDescription,
 } from '../autocomplete/body_completer';
+import { Endpoint } from '../autocomplete/types';
```
where is this `../autocomplete/types`? did we define it somewhere?

curq replied:
Sorry, I forgot to mention that this PR should be merged after https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148, where this file was created.

ananzh commented on src/plugins/console/public/lib/osd/api.ts:
```diff
@@ -29,12 +29,14 @@
   compileBodyDescription,
 } from '../autocomplete/body_completer';
+import { Endpoint } from '../autocomplete/types';
+import { ParametrizedComponentFactories } from './osd';
```
if a type file is defined, should move `ParametrizedComponentFactories` in a single type file

curq replied:
Will do

ananzh commented on src/plugins/console/public/lib/osd/api.ts:
```diff
@@ -29,12 +29,14 @@
  */
 
 import _ from 'lodash';
```
same lodash issue

ananzh commented on src/plugins/console/public/lib/osd/api.ts:
```diff
@@ -44,41 +46,50 @@ import {
-  cls.getGlobalAutocompleteComponents = function (term, throwOnMissing) {
-    const result = this.globalRules[term];
+  getGlobalAutocompleteComponents(term: string, throwOnMissing?: boolean) {
+    const result: SharedComponent[] = this.globalRules[term];
```
`globalRules` is defined as `Record<string, any>`. `Record<string, any>` implies that the values in `globalRules` could be of any type, so we can't directly assert that this.globalRules[term] is of type SharedComponent[], right?

curq replied:
Yes, that's right. I referred to the type defined in file that already was in TS.
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/487cd363abd91559369bb9c8fc7771175265bd0d/src/plugins/console/server/services/spec_definitions_service.ts#L43-L45

ananzh commented on typings/index.d.ts:
```diff
@@ -46,6 +46,12 @@ declare module '*.svg' {
   export default content;
 }
 
+declare module '*.txt' {
```
curious, why add this?

curq replied:
The ```curl_parsing.test``` imports txt file for the testing purposes and TS required module declaration

ananzh commented on src/plugins/console/public/lib/curl_parsing/curl.ts:
```diff
@@ -79,12 +79,12 @@ export function parseCURL(text) {
       return false;
     }
-    line = lines.shift().replace(/[\r\n]+/g, '\n') + '\n';
+    line = lines.shift()?.replace(/[\r\n]+/g, '\n') + '\n';
```
so line could be undefined?

curq replied:
.shift only returns undefined when lines array is empty and this condition was tested in lines above ```(lines.length === 0)```, I should change it to ! instead.

ananzh commented on src/plugins/console/public/lib/osd/api.ts:
```diff
@@ -44,41 +46,50 @@ import {
 
-  cls.getGlobalAutocompleteComponents = function (term, throwOnMissing) {
-    const result = this.globalRules[term];
+  getGlobalAutocompleteComponents(term: string, throwOnMissing?: boolean) {
```
why only throwOnMissing has `?`

curq replied:
I looked through the code and whenever this method used, the term was always present while throwOnMissing was optional.

curq replied:
Here is the example, where throwOnMissing is optional
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d94d7cb683aa76878bba5120cb8130add147d7ad/src/plugins/console/public/lib/autocomplete/autocomplete.ts#L883
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d94d7cb683aa76878bba5120cb8130add147d7ad/src/plugins/console/public/lib/autocomplete/body_completer.ts#L127

ananzh reviewed (commented):
I think we need to verify all the `?` usage.

curq commented on src/plugins/console/public/lib/osd/osd.ts:
```diff
@@ -172,7 +207,7 @@ function loadApisFromJson(
-export function setActiveApi(api) {
+export function setActiveApi(api?: Api) {
   if (!api) {
     $.ajax({
```
1. I will update it alongside of lodash/jquery removal. I planned to make separate PRs for TS and other refactoring to make reviewing easier.
2. Yes, there is code where the functions is called with no arguments https://github.com/opensearch-project/OpenSearch-Dashboards/blob/487cd363abd91559369bb9c8fc7771175265bd0d/src/plugins/console/public/lib/osd/osd.js#L196-L198

ananzh replied:
1. cool. thanks.
2. nice catch.

curq commented on src/plugins/console/public/lib/osd/api.ts:
```diff
@@ -44,41 +46,50 @@ import {
 
-  cls.addEndpointDescription = function (endpoint, description) {
-    const copiedDescription = {};
+  addEndpointDescription(endpoint: string, description?: Record<string, any>) {
```
I think the ? is likely redundant here, I will review the ? and will remove if this is the case, thank for the catch.

curq replied:
I removed this redundant `?`. I've gone through the other usages of `?` and we do need them.

curq commented on src/plugins/console/public/lib/osd/osd.ts:
```diff
@@ -124,37 +159,37 @@ export function getUnmatchedEndpointComponents() {
-  json,
-  urlParametrizedComponentFactories,
-  bodyParametrizedComponentFactories
+  json: Record<string, any>,
```
In all the cases where this function was used the json was always present, so I assumed it was required,

curq commented:
@ananzh Thank you for the review! I responded to your comments. I forgot to initially mention the order of merging/review, it should first be https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4008, then https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148 and only after that this one. The reason for that is that they have types on which this PR is dependent. So I think it would be better to return to this one after they are finished, because the changes there may also affect this PR.

Regarding lodash/jquey, I was planning to remove it in separate PR in order to not mix new TS logic and lodash/jquery refactoring for better review experience. If you think it is better to do it all in this PR, I'm ready to do it there, just let me know.

ananzh commented on src/plugins/console/public/lib/osd/osd.ts:
```diff
@@ -28,6 +28,8 @@
  */
 
+import $ from 'jquery';
+import _ from 'lodash';
```
sure. I am okay with that.

abbyhu2000 commented:
The pre-req for this PR are merged: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4008 and https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4148

curq commented:
@abbyhu2000 Thanks for the heads up! I've addressed all the comments and updated my PR. It's now ready for review. @ananzh Could you please check it out?

joshuarrrr reviewed (approved):
LGTM
